awx/main/models/events.py

```diff
@@ -0,0 +1,685 @@
+        #
+        # In the event of any issue, throw it out, and Django will just save
+        # the current time.
+        try:
+            if not isinstance(kwargs['created'], datetime.datetime):
+                kwargs['created'] = parse_datetime(kwargs['created'])
+            if not kwargs['created'].tzinfo:
+                kwargs['created'] = kwargs['created'].replace(tzinfo=utc)
+        except (KeyError, ValueError):
+            kwargs.pop('created', None)
+
+        # events that not jobs/adhoc (generic) are just event=verbose
+        if 'job_id' not in kwargs and 'ad_hoc_command_id' not in kwargs:
+            kwargs['event'] = 'verbose'
+
+        # Sanity check: Don't honor keys that we don't recognize.
+        valid_keys = {'ad_hoc_command_id', 'unified_job_id', 'event',
+                      'event_data', 'created', 'counter', 'uuid', 'stdout',
+                      'start_line', 'end_line', 'verbosity'}
+        for key in kwargs.keys():
+            if key not in valid_keys:
+                kwargs.pop(key)
+
+        return self.objects.create(**kwargs)
+
+
+class AdHocCommandEvent(BaseCommandEvent):
+
+    class Meta:
+        app_label = 'main'
+        ordering = ('-pk',)
+        index_together = [
+            ('ad_hoc_command', 'event'),
+            ('ad_hoc_command', 'uuid'),
+            ('ad_hoc_command', 'start_line'),
+            ('ad_hoc_command', 'end_line'),
+        ]
+
+    ad_hoc_command = models.ForeignKey(
+        'AdHocCommand',
+        related_name='ad_hoc_command_events',
+        on_delete=models.CASCADE,
+        editable=False,
+    )
+    host = models.ForeignKey(
+        'Host',
+        related_name='ad_hoc_command_events',
+        null=True,
+        default=None,
+        on_delete=models.SET_NULL,
+        editable=False,
+    )
+    host_name = models.CharField(
+        max_length=1024,
+        default='',
+        editable=False,
+    )
+
+    def save(self, *args, **kwargs):
+        update_fields = kwargs.get('update_fields', [])
+        # If update_fields has been specified, add our field names to it,
+        # if it hasn't been specified, then we're just doing a normal save.
+        self.host_name = self.event_data.get('host', '').strip()
+        if 'host_name' not in update_fields:
+            update_fields.append('host_name')
+        if not self.host_id and self.host_name:
+            host_qs = self.ad_hoc_command.inventory.hosts.filter(name=self.host_name)
+            try:
+                host_id = host_qs.only('id').values_list('id', flat=True)
+                if host_id.exists():
+                    self.host_id = host_id[0]
+                    if 'host_id' not in update_fields:
+                        update_fields.append('host_id')
+            except (IndexError, AttributeError):
+                pass
+        super(AdHocCommandEvent, self).save(*args, **kwargs)
+
+
+class GenericCommandEvent(BaseCommandEvent):
+
+    unified_job = models.ForeignKey(
```

Comment:
We should probably index some of this in the same way as https://github.com/ansible/awx/pull/833/files#diff-9b1e5b80fcb01fab3e7a0ece4e371a50R631; I expect it'll be queried similarly by the UI.